pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -118,24 +131,73 @@ func ValidateContainerConcurrency(cc RevisionContainerConcurrencyType, cm Revisi
 	return nil
 }
 
-func validateContainer(container corev1.Container) *apis.FieldError {
+func validateVolume(volume corev1.Volume) *apis.FieldError {
+	var errs *apis.FieldError
+	if volume.Name == "" {
+		errs = apis.ErrMissingField("name")
+	}
+	vs := volume.VolumeSource
+	switch {
+	case vs.Secret != nil, vs.ConfigMap != nil:
+		// These are fine.
+	default:
+		errs = errs.Also(apis.ErrMissingOneOf("secret", "configMap"))
+	}
+	return errs
+}
+
+func validateContainer(container corev1.Container, volumes sets.String) *apis.FieldError {
 	if equality.Semantic.DeepEqual(container, corev1.Container{}) {
 		return apis.ErrMissingField(apis.CurrentField)
 	}
+
+	// Check that volume mounts match names in "volumes", that "volumes" has 100%
+	// coverage, and the field restrictions.
+	seen := sets.NewString()
+	var errs *apis.FieldError
+	for i, vm := range container.VolumeMounts {
+		// This effectively checks that Name is non-empty because Volume name must be non-empty.
+		if !volumes.Has(vm.Name) {
+			errs = errs.Also(
+				apis.ErrInvalidValue(vm.Name, "name").ViaFieldIndex("volumeMounts", i))
```

Comment:
This error types seems misleading here. It is checking that the name is non-empty for volumeMounts and that there is a match is volumes. 

I think it makes sense for empty, but I am not sure I would easily get that I have a mismatch from an invalidValue error. Can we get a more specific error here?